### ISSUE TITLE
Progressive narrowing of ivy-occur buffer

### ISSUE THREAD

**Issue by lsp-ableton** [abo-abo/swiper#2347]
I have an idea for an improvement to `ivy-occur`. I'll take a crack at implementing it if people think it's a good idea, but I wanted to check first. 

My proposal is that if you are searching inside an 'ivy-occur' buffer with swiper and then call `ivy-occur` again, the resulting buffer would filter candidates from the first `ivy-occur` buffer instead of referencing the original.

The logic behind this is that sometimes it's useful to make a broad search, and then after you look through the candidates, you narrow it down further. In the end you may want to use wgrep to edit the occurrences. You could probably do the same sort of thing with a more complicated query-replace/ex command in the original buffer, but I like the idea of progressively narrowing the results with swiper and then doing the edit.

Thoughts?

**Comment by Alexander-Shukaev** [comment]
I was under impression that cascaded candidate filtering/narrowing is already part of `ivy`, where one could make the result candidate list of the first filtering session to be an input of the second filtering session and so on.  Maybe somebody can comment how to turn on or trigger this feature?

**Comment by leungbk** [comment]
> I was under impression that cascaded candidate filtering/narrowing is already part of `ivy`, where one could make the result candidate list of the first filtering session to be an input of the second filtering session and so on. Maybe somebody can comment how to turn on or trigger this feature?

`ivy-restrict-to-matches` works only within the ivy minibuffer, and I'm not aware of any analog for `ivy-occur` buffers.

**Comment by Alexander-Shukaev** [comment]
Right, and what I mean is that doesn't it look superior to what @lsp-ableton wants?

By the way, `ivy-restrict-to-matches` does not work with `dynamic-collection t` (try `counsel-rg` for example), and if called in that case, will actually break the input and navigation.  It's definitely a bug and `ivy-restrict-to-matches` should do nothing or throw error when called during `dynamic-collection t`.

**Comment by leungbk** [comment]
> Right, and what I mean is that doesn't it look superior to what @lsp-ableton wants?

I'm not sure if it's possible to use it straight from an old `ivy-occur` buffer though.

> By the way, `ivy-restrict-to-matches` does not work with `dynamic-collection t` (try `counsel-rg` for example), and if called in that case, will actually break the input and navigation. It's definitely a bug and `ivy-restrict-to-matches` should do nothing or throw error when called during `dynamic-collection t`.

It works for me with `counsel-rg` (https://github.com/abo-abo/swiper/commit/230137be3ea4cce20bed0689481c4fc4394c3867). It doesn't work with `swiper-isearch` though.

**Comment by Alexander-Shukaev** [comment]
> I'm not sure if it's possible to use it straight from an old `ivy-occur` buffer though.

Surely not, and the point is to call `ivy-occur` in the very end, when multi-stage narrowing is done.

> It works for me with `counsel-rg` (230137b). It doesn't work with `swiper-isearch` though.

Did you wait until `rg` finished populating candidates in the minibuffer before `ivy-restrict-to-matches`?

**Comment by lsp-ableton** [comment]
> Surely not, and the point is to call `ivy-occur` in the very end, when multi-stage narrowing is done.
> 

`ivy-restrict-to-matches` is nice and quick for simpler narrowing, but repeatedly narrowing an `ivy-occur` buffer covers more use cases.

1. You can manually delete candidates from an `ivy-occur` buffer
2. You can interactively narrow based on the file path when using `counsel-ag` or `counsel-rg`
3. You have a dedicated buffer for candidates, so you could investigate them in more detail before narrowing further

I don't really see a reason not to do it, because I don't think there's any chance that someone searching an occur buffer really wants the candidates that result from doing another occur to refer back to the occur buffer rather than the original results.

**Comment by leungbk** [comment]
> Did you wait until `rg` finished populating candidates in the minibuffer before `ivy-restrict-to-matches`?

Yes, this is with the latest master branch.
